### DISCUSSION TITLE
How to limit number of threads in threadpool

### DISCUSSION THREAD

**Discussion by cyrusbehr** [fastapi/fastapi#6361] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, status, File

import numpy as np
from colorama import Fore
from colorama import Style
import os

import time
from concurrent.futures import ThreadPoolExecutor
import asyncio


app = FastAPI()


@app.on_event("startup")
def set_default_executor():
    loop = asyncio.get_running_loop()
    loop.set_default_executor(
        ThreadPoolExecutor(max_workers=2)
    )

    print("The number of workers has been set to 2")


@app.post('/cpu-bound-task')
def feature_vector(image: bytes = File(...)):
    nparr = np.fromstring(image, np.uint8)

    # Some CPU bound machine learning task here....

# pip install fastapi uvicorn python-multipart
#
# run with
#
# uvicorn test_fast_api:app --workers 1 --port 8080 --reload
```

### Description

I am writing a server which runs CPU bound tasks (machine learning). 

I'd like to make use of parallelism to maximize CPU resource.
Due to the machine learning library which will be called, I cannot spawn multiple processes, and instead need to use mutlthreading (the library releases the python GIL in the binding layer so we can take advantage of CPU resources even with python multithreading restrictions). 

Based on what I read, when you define a synchronous endpoint, FastAPI makes use of a threadpool. This is the desired behaviour I want. However, I need to be able to limit the number of threads in the threadpool - the reason being that my machine learning library uses 8 threads for inference, so if too many concurrent requests are recieved, having too many threads in the threadpool will cause timeslicing and slow things down significantly). 

I already tried this solution [here](https://github.com/tiangolo/fastapi/issues/603#issuecomment-910055647), but it did not limit the number of threads that were launched (the number of threads matched the number of concurrent requests). Therefore, how do I limit the number of threads in the threadpool?  

### Operating System

Linux

### Operating System Details

Ubuntu 18.04

### FastAPI Version

0.75.1

### Python Version

Python 3.8.8

**Comment by jgould22** [comment, marked as answer]
Usually the common approach here is to use something like Celery (or any other similar library) to keep separate the web layer from the CPU intensive tasks. That way you have more control and your web server is never in a situation where it cannot respond, even to say it is overloaded or queue things.

**Comment by cyrusbehr** [comment]
@jgould22 thank you for the suggestion, I agree that it may be best to use a queue / message broker.
